packages/flutter/lib/src/widgets/router.dart

```diff
@@ -1032,27 +1035,37 @@ class _BackButtonListenerState extends State<BackButtonListener> {
 
   @override
   void didChangeDependencies() {
+    _updateCallback(widget.onBackPressed);
     super.didChangeDependencies();
-    dispatcher?.removeCallback(widget.onBackPressed);
-
-    final BackButtonDispatcher? rootBackDispatcher = Router.of(context).backButtonDispatcher;
-    assert(rootBackDispatcher != null, 'make sure to have a Router object with a backButtonDispatcher registered');
-
-    dispatcher = rootBackDispatcher!.createChildBackButtonDispatcher()
-      ..addCallback(widget.onBackPressed)
-      ..takePriority();
   }
 
   @override
-  Widget build(BuildContext context) {
-    return widget.child;
+  void didUpdateWidget(covariant BackButtonListener oldWidget) {
+    _updateCallback(oldWidget.onBackPressed);
```

Comment:
for did update widget, you don't want to recreate the dispatcher, it will spend time clearing up the back dispatcher chain. We should just do something like
```dart
if (oldWidget.onBackPressed != widget.onBackPressed) {
  dispatcher?.removeCallback(oldWidget.onBackPressed);
  dispatcher?.addCallback(widget.onBackPressed);
}
```